src/ty_check/mod.rs

```diff
@@ -2314,110 +3577,214 @@ impl TyChecker {
+        let trait_def = self.gl_ctx.trait_ty_by_name(name)?;
+        self.dty_well_formed(kind_ctx, ty_ctx, exec, &constraint.param)?;
+        Self::check_args_have_correct_kinds(&trait_def.generic_params, generics)?;
+        iter_TyResult_to_TyResult!(generics
+            .iter()
+            .map(|gen| self.arg_kinded_well_formed(kind_ctx, ty_ctx, exec, gen)))
+    }
 }
 
-pub fn proj_elem_ty(ty: &Ty, i: usize) -> TyResult<Ty> {
-    match &ty.ty {
-        TyKind::Data(DataTy {
-            dty: DataTyKind::Tuple(dtys),
-            ..
-        }) => match dtys.get(i) {
+pub fn proj_elem_ty(ty: &Ty, proj: &ProjEntry) -> TyResult<Ty> {
+    match (&ty.ty, proj) {
+        (
+            TyKind::Data(DataTy {
+                dty: DataTyKind::Tuple(dtys),
+                ..
+            }),
+            ProjEntry::TupleAccess(i),
+        ) => match dtys.get(*i) {
             Some(dty) => Ok(Ty::new(TyKind::Data(dty.clone()))),
-            None => Err(TyError::String(format!(
+            None => Err(TyError::IllegalProjection(format!(
                 "Cannot project element `{}` from tuple with {} elements.",
                 i,
                 dtys.len()
             ))),
         },
-        _ => Err(TyError::String(
-            "Cannot project from non tuple type.".to_string(),
-        )),
+        (
+            TyKind::Data(DataTy {
+                dty: DataTyKind::Struct(struct_ty),
+                ..
+            }),
+            ProjEntry::StructAccess(attr_name),
+        ) => match struct_ty.get_ty(attr_name) {
+            Some(ty) => Ok(Ty::new(TyKind::Data(ty.clone()))),
+            None => Err(TyError::IllegalProjection(format!(
+                "Cannot project element `{}` from struct {}.",
+                attr_name, struct_ty.name
+            ))),
+        },
+        _ => Err(TyError::IllegalProjection(format!(
+            "Unexpected type or projection access. Found access: {:?} on ty {:#?}",
+            proj, ty
+        ))),
+    }
+}
+
+fn is_ty_copyable(
+    ty: &Ty,
+    constraint_env: &ConstraintEnv,
+    implicit_ident_cons: &mut IdentConstraints,
+) -> bool {
+    match &ty.ty {
+        TyKind::Data(dty) => is_dty_copyable(dty, constraint_env, implicit_ident_cons),
+        TyKind::Fn(_, _, _) => true,
+        TyKind::Ident(_) => false,
+        TyKind::Dead(_) => panic!(
+            "This case is not expected to mean anything.\
+            The type is dead. There is nothing we can do with it."
+        ),
+    }
+}
+
+fn is_dty_copyable(
+    dty: &DataTy,
+    constraint_env: &ConstraintEnv,
+    implicit_ident_cons: &mut IdentConstraints,
+) -> bool {
+    use DataTyKind::*;
+
+    let is_dty_copy_constraint = Constraint {
+        param: dty.clone(),
+        trait_bound: copy_trait(),
+    };
+
+    match &dty.dty {
+        // FIXME thread hierarchies and their splits should be non-copyable!
+        ThreadHierchy(_) => true,
+        SplitThreadHierchy(_, _) => true,
```

Comment:
TODO Die Syntax dafür im Parser nachschauen und dann die Überprüfung in die std-lib packen